File: MATLAB/RGB to HSI.m
function rgb_to_hsi_demo()
    RGB = imread('Enter your image path here'); 
    HSI = rgb2hsi(RGB);
    figure('Position', [100, 100, 1200, 800]);
    subplot(2, 3, 1);
    imshow(RGB, 'InitialMagnification', 'fit'); 
    title('Original RGB Image', 'FontSize', 12);
    subplot(2, 3, 2), imshow(HSI(:,:,1), 'InitialMagnification', 'fit'), title('Hue', 'FontSize', 12);
    subplot(2, 3, 3), imshow(HSI(:,:,2), 'InitialMagnification', 'fit'), title('Saturation', 'FontSize', 12);
    subplot(2, 3, 4), imshow(HSI(:,:,3), 'InitialMagnification', 'fit'), title('Intensity', 'FontSize', 12);
    combined_hsi = im2uint8(HSI); 
    subplot(2, 3, 5), imshow(combined_hsi, 'InitialMagnification', 'fit'), title('Combined HSI', 'FontSize', 12);
    subplot(2, 3, 6), axis off; 
end
function HSI = rgb2hsi(RGB)
    RGB = double(RGB) / 255;
    R = RGB(:,:,1);
    G = RGB(:,:,2);
    B = RGB(:,:,3);
    I = (R + G + B) / 3;
    num = min(R, min(G, B));
    den = I + eps; 
    S = 1 - (num ./ den); % S = 1 - (min(R,G,B) / I)
    denom = sqrt((R - G).^2 + (R - B) .* (G - B) + eps);
    theta = acos(0.5 * ((R - G) + (R - B)) ./ denom); 
    H = zeros(size(I)); % Initialize Hue matrix
    H(B <= G) = theta(B <= G);               
    H(B > G) = 2 * pi - theta(B > G);      
    H = H / (2 * pi);
    HSI = cat(3, H, S, I);
end

File: Python/RGB to HSI.py
import numpy as np
import matplotlib.pyplot as plt
import imageio.v2 as imageio

def rgb2hsi(rgb_img):
    rgb = rgb_img.astype(np.float32) / 255
    R = rgb[:, :, 0]
    G = rgb[:, :, 1]
    B = rgb[:, :, 2]
    I = (R + G + B) / 3
    min_rgb = np.minimum(np.minimum(R, G), B)
    S = 1 - (min_rgb / (I + 1e-8))
    num = 0.5 * ((R - G) + (R - B))
    den = np.sqrt((R - G) ** 2 + (R - B) * (G - B)) + 1e-8
    theta = np.arccos(num / den)
    H = np.zeros_like(I)
    H[B <= G] = theta[B <= G]
    H[B > G] = 2 * np.pi - theta[B > G]
    H = H / (2 * np.pi)
    HSI = np.stack((H, S, I), axis=-1)
    return HSI

img = imageio.imread('Enter your image path here')
hsi = rgb2hsi(img)
hsi_uint8 = (hsi * 255).astype(np.uint8)

plt.figure(figsize=(12, 8))
plt.subplot(2, 3, 1)
plt.imshow(img)
plt.title('Original RGB Image', fontsize=12)
plt.axis('off')

plt.subplot(2, 3, 2)
plt.imshow(hsi[:, :, 0], cmap='hsv')
plt.title('Hue', fontsize=12)
plt.axis('off')

plt.subplot(2, 3, 3)
plt.imshow(hsi[:, :, 1], cmap='gray')
plt.title('Saturation', fontsize=12)
plt.axis('off')

plt.subplot(2, 3, 4)
plt.imshow(hsi[:, :, 2], cmap='gray')
plt.title('Intensity', fontsize=12)
plt.axis('off')

plt.subplot(2, 3, 5)
plt.imshow(hsi_uint8)
plt.title('Combined HSI', fontsize=12)
plt.axis('off')

plt.subplot(2, 3, 6)
plt.axis('off')

plt.tight_layout()
plt.show()
